[RDY] test: migrate legacy eval test.

lucc opened the pull request:
The original test file did a `:wq` somewhere in the middle.  I "ressurected" the tests that where in test_eval.in after this point but I had to make up the results (they where not in test_eval.ok).  The commit for this is the last one (currently 60b1139).

lucc committed:
tests: Migrate legacy eval test.

lucc committed:
tests: Debug converted eval test.

lucc committed:
tests: Split converted eval test into it() blocks.

lucc committed:
tests: Resurrect unused part of legacy eval test.

The old test_eval.in file did a `:wq` on line 168.  The following 60 lines
where not used and there was no expected output for them in test_eval.ok.
This test code is now used in several test cases in the new test file.

lucc committed:
tests: Comments and typos after review.

Helped-by: Justin M. Keyes <justinkz@gmail.com>

lucc commented:
This is now RFC. The resurrection commit is now 91b9f70.

justinmk commented on test/functional/legacy/eval_spec.lua:
```diff
@@ -0,0 +1,667 @@
+local clear, execute, expect = helpers.clear, helpers.execute, helpers.expect
+local eq, eval, wait, write_file = helpers.eq, helpers.eval, helpers.wait, helpers.write_file
+
+describe('various eval features', function()
```
just `'eval'` is enough

justinmk commented on test/functional/legacy/eval_spec.lua:
```diff
@@ -0,0 +1,667 @@
+    os.remove('test_eval_setup.vim')
+  end)
+
+  it('let tests', function()
```
just `':let'` is enough.

justinmk commented on test/functional/legacy/eval_spec.lua:
```diff
@@ -0,0 +1,667 @@
+    eq('Vim(function):E128: Function name must start with a capital or "s:": g:test()', eval('tmp'))
+  end)
+
+  it('Function name folowed by #', function()
```
typo

justinmk commented:
@lucc Just curious, did we somehow delete these tests at some point (not sure what you mean by resurrect)?

This LGTM, after squash I will do a fuzz-test locally then merge.

lucc committed:
tests: Migrate recent patches of legacy eval test.

As the work of this migration did take quite some time, some changes where
made in master after this branch was started.  These are ported to the new
test file.

lucc commented:
I added three more commits. One to apply the comments above. The next two are copying over changes that where made to the original eval test since I started this branch.

About the resurrect question: The original test [exits at line 168](https://github.com/neovim/neovim/blob/master/src/nvim/testdir/test_eval.in#L168). The code afterwards is not used and there is no corresponding data in test_eval.ok (the last part in the ok file is about setreg errors).

I used this previously (in vim and nvim) unused test code and "resurrected" it to [lines 599-697 in the new test](https://github.com/lucc/neovim/blob/legacy/eval/test/functional/legacy/eval_spec.lua#L599-L697). I had to think about the expected results myself, please double check them if you have not done so already.

I will now start to squash this but I will put a backup branch with all these WIP commits up on my account so that it can still be discussed if need be.

lucc commented:
The backup of the WIP commits is [here](https://github.com/lucc/neovim/tree/legacy/eval.presquash.bak).

lucc committed:
tests: Remove legacy test files for eval test.

lucc commented:
I consider it RDY.

bfredl commented:
Is there a purpose here to really test the system clipboard integration, or just to test eval stuff that happens to involve `"*` register? If the later, the clipboard can be virtualized using
```
    execute('let &rtp = "test/functional/fixtures,".&rtp')
```
(should be placed right after a `clear()` )

lucc commented:
@bfredl I did not spend a thought on this. I just tried to migrate *everything* that was in the legacy test. The old eval test tried to test the (real) clipbord integration as I read from the test file and the hg log of the vim repository (`hg log -v --patch src/testdir/test_eval*`).

We could replace the `if has_clipboard()` and `pending()` stuff by just the virtualized clipboard. Actually I think that this test does not test if the yanked text ends up in the system clipboard. It just tests if we can set and get from the `*` register, no check outside of the nvim process is made if the system clipboard was set. I hope that a actual test for a change in the system clipboard is made in `test/functional/clipboard/*`.

justinmk commented on test/functional/legacy/eval_spec.lua:
```diff
@@ -0,0 +1,699 @@
+  before_each(clear)
+  teardown(function()
+    os.remove('test.out')
+    os.remove('test_eval_func.vim')
```
I do not see where this is ever written

justinmk commented on test/functional/legacy/eval_spec.lua:
```diff
@@ -0,0 +1,699 @@
+      =: type v; value: abc/]].."\x00 (['abc/\x00"..[[']), expr: "abc/]]..'\x00'..[[" (['"abc/]]..'\x00'..[["'])]])
+  end)
+
+  -- If the clipboard feature is not available the next test will be sciped.
```
sciped => skipped. But the comment is actually not needed.

justinmk commented on test/functional/legacy/eval_spec.lua:
```diff
@@ -0,0 +1,699 @@
+      endfun
+
+      fun AppendRegParts(reg, type, cont, strcont, cont1, strcont1)
+	  call append('$', printf('%s: type %s; value: %s (%s), expr: %s (%s)', a:reg, a:type, a:cont, a:strcont, a:cont1, a:strcont1))
```
convert tab indent to spaces

justinmk commented:
Thanks @lucc
